File: src/go/signalbird/management.go
package signalbird

import "context"

// Management — Yönetim istemcisi.
//
// Müşterinin panelde tıklayarak yaptığı her şeyi kodla yapar: Telsiz projesi
// ve kanalı, sohbet gelen kutusu, uygulama kaydı ve cihaz listesi.
//
// Bu ADMIN yüzeyi DEĞİLDİR: anahtar tek bir takıma bağlıdır ve yalnız o
// takımın kayıtlarına dokunur; başka takımın kaydı 404 döner.
//
// Sözleşme: docs/CONTRACT.md § 10
type Management struct {
	http *transport
}

// NewManagement kurar. Gönderim ile aynı anahtar ailesini kullanır (sb_…) ama
// radio:*, chat:*, apps:* scope'larını ister.
func NewManagement(config KeyConfig) (*Management, error) {
	http, err := newKeyTransport(config)
	if err != nil {
		return nil, err
	}

	return &Management{http: http}, nil
}

// ── Telsiz: projeler ───────────────────────────────────────────────────

func (m *Management) RadioSummary(ctx context.Context) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/radio/summary", nil, nil)
}

func (m *Management) RadioEvents(ctx context.Context, query Query) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/radio/events", nil, query)
}

// ── Modül anahtarları ──────────────────────────────────────────────────
//
// Telsiz projesi/kanalı ve uygulama kaydı 1 Eyl 2026'da kaldırıldı
// (../signalbird.api/docs/KEY_ARCHITECTURE_2026-09-01.md §3). Beş modül
// (logger, email, sms, push, chat) aynı gövdeyi kullanır.

func (m *Management) ListModuleKeys(ctx context.Context, module string, query Query) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/modules/"+seg(module)+"/keys", nil, query)
}

func (m *Management) GetModuleKey(ctx context.Context, module string, id any) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/modules/"+seg(module)+"/keys/"+seg(id), nil, nil)
}

// CreateModuleKey — `key` verilmezse başlıktan üretilir; çakışırsa sayı eklenir.
func (m *Management) CreateModuleKey(ctx context.Context, module string, input any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/modules/"+seg(module)+"/keys", input, nil)
}

// UpdateModuleKey — `key` DEĞİŞTİRİLEBİLİR: eski ad 30 gün daha kabul edilir.
func (m *Management) UpdateModuleKey(ctx context.Context, module string, id any, input any) (Result, error) {
	return m.http.request(ctx, "PATCH", "/v1/modules/"+seg(module)+"/keys/"+seg(id), input, nil)
}

func (m *Management) DeleteModuleKey(ctx context.Context, module string, id any) (Result, error) {
	return m.http.request(ctx, "DELETE", "/v1/modules/"+seg(module)+"/keys/"+seg(id), nil, nil)
}

// ListModuleKeyDevices — push kanalının cihazları; token MASKELİ döner.
func (m *Management) ListModuleKeyDevices(ctx context.Context, module string, id any, query Query) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/modules/"+seg(module)+"/keys/"+seg(id)+"/devices", nil, query)
}

// ── Sohbet: gelen kutusu ───────────────────────────────────────────────

func (m *Management) ChatSummary(ctx context.Context) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/chat/summary", nil, nil)
}

func (m *Management) ChatUpdates(ctx context.Context) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/chat/updates", nil, nil)
}

func (m *Management) ListConversations(ctx context.Context, query Query) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/chat/conversations", nil, query)
}

func (m *Management) GetConversation(ctx context.Context, id string) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/chat/conversations/"+seg(id), nil, nil)
}

// ListConversationMessages — after imleci cm_… mesaj kimliğidir.
func (m *Management) ListConversationMessages(ctx context.Context, id string, query Query) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/chat/conversations/"+seg(id)+"/messages", nil, query)
}

// StartConversation — proaktif sohbet; ziyaretçi yazmadan ajan başlatır.
func (m *Management) StartConversation(ctx context.Context, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/chat/conversations", input, nil)
}

func (m *Management) UpdateConversation(ctx context.Context, id string, input map[string]any) (Result, error) {
	return m.http.request(ctx, "PATCH", "/v1/chat/conversations/"+seg(id), input, nil)
}

func (m *Management) SetConversationStatus(ctx context.Context, id, status string) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/chat/conversations/"+seg(id)+"/status", map[string]any{"status": status}, nil)
}

// AssignConversation — userID 0 ise anahtarın sahibine atanır.
func (m *Management) AssignConversation(ctx context.Context, id string, userID int) (Result, error) {
	var value any

	if userID > 0 {
		value = userID
	}

	return m.http.request(ctx, "POST", "/v1/chat/conversations/"+seg(id)+"/assign", map[string]any{"user_id": value}, nil)
}

func (m *Management) ReadConversation(ctx context.Context, id, lastMessageID string) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/chat/conversations/"+seg(id)+"/read", map[string]any{"last_message_id": lastMessageID}, nil)
}

func (m *Management) SetTyping(ctx context.Context, id string, isTyping bool) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/chat/conversations/"+seg(id)+"/typing", map[string]any{"is_typing": isTyping}, nil)
}

// Reply — is_internal true ise iç nottur ve ziyaretçiye ASLA gitmez.
func (m *Management) Reply(ctx context.Context, id string, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/chat/conversations/"+seg(id)+"/messages", input, nil)
}

func (m *Management) EditChatMessage(ctx context.Context, id, messageID, body string) (Result, error) {
	return m.http.request(ctx, "PATCH", "/v1/chat/conversations/"+seg(id)+"/messages/"+seg(messageID), map[string]any{"body": body}, nil)
}

func (m *Management) DeleteChatMessage(ctx context.Context, id, messageID string) (Result, error) {
	return m.http.request(ctx, "DELETE", "/v1/chat/conversations/"+seg(id)+"/messages/"+seg(messageID), nil, nil)
}

// ReactToChatMessage — aynı emoji ikinci kez gönderilirse tepki kaldırılır.
func (m *Management) ReactToChatMessage(ctx context.Context, id, messageID, emoji string) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/chat/conversations/"+seg(id)+"/messages/"+seg(messageID)+"/reactions", map[string]any{"emoji": emoji}, nil)
}

// ── Sohbet: ziyaretçi ve hazır yanıtlar ────────────────────────────────

func (m *Management) GetVisitor(ctx context.Context, id string) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/chat/visitors/"+seg(id), nil, nil)
}

func (m *Management) UpdateVisitor(ctx context.Context, id string, input map[string]any) (Result, error) {
	return m.http.request(ctx, "PATCH", "/v1/chat/visitors/"+seg(id), input, nil)
}

func (m *Management) BanVisitor(ctx context.Context, id string) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/chat/visitors/"+seg(id)+"/ban", nil, nil)
}

func (m *Management) ListCannedReplies(ctx context.Context) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/chat/canned-replies", nil, nil)
}

func (m *Management) CreateCannedReply(ctx context.Context, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/chat/canned-replies", input, nil)
}

func (m *Management) UpdateCannedReply(ctx context.Context, id any, input map[string]any) (Result, error) {
	return m.http.request(ctx, "PATCH", "/v1/chat/canned-replies/"+seg(id), input, nil)
}

func (m *Management) DeleteCannedReply(ctx context.Context, id any) (Result, error) {
	return m.http.request(ctx, "DELETE", "/v1/chat/canned-replies/"+seg(id), nil, nil)
}

// ── Sohbet: tetikleyiciler ───────────────────────────────────────────────
// "Şu olduğunda şunu yap." Kural KAYITTA durur, kodda değil.

func (m *Management) ListChatTriggers(ctx context.Context) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/chat/triggers", nil, nil)
}

func (m *Management) CreateChatTrigger(ctx context.Context, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/chat/triggers", input, nil)
}

func (m *Management) UpdateChatTrigger(ctx context.Context, id any, input map[string]any) (Result, error) {
	return m.http.request(ctx, "PATCH", "/v1/chat/triggers/"+seg(id), input, nil)
}

func (m *Management) DeleteChatTrigger(ctx context.Context, id any) (Result, error) {
	return m.http.request(ctx, "DELETE", "/v1/chat/triggers/"+seg(id), nil, nil)
}

// ── Sohbet: rapor ────────────────────────────────────────────────────────

// ChatReport yanıt/çözüm süresi, memnuniyet ve ajan kırılımını döner.
// Veri yoksa süreler null döner — 0 DEĞİL.
func (m *Management) ChatReport(ctx context.Context, rng string) (Result, error) {
	if rng == "" {
		rng = "30d"
	}

	return m.http.request(ctx, "GET", "/v1/chat/reports", nil, map[string]any{"range": rng})
}

// Uygulama uçları KALDIRILDI (1 Eyl 2026): sohbet ve push birer modül
// anahtarıdır — ListModuleKeys(ctx, "chat", nil).

func (m *Management) EmbedToken(ctx context.Context, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/embed/tokens", input, nil)
}

File: src/python/signalbird/management.py
"""Yönetim (Management) istemcisi — sunucu tarafı.

Müşterinin panelde tıklayarak yaptığı her şeyi kodla yapar: Telsiz projesi ve
kanalı açar, olay akışını okur, sohbet gelen kutusunu işler, uygulama kaydı ve
cihaz listesi yönetir.

Bu ADMIN yüzeyi DEĞİLDİR: anahtar tek bir takıma bağlıdır ve yalnız o takımın
kayıtlarına dokunur.

Sözleşme: docs/CONTRACT.md § 10
"""

from __future__ import annotations

from typing import Any, Mapping, Optional

from ._http import DEFAULT_BASE_URL, Result, SignalbirdError, Transport, seg


class SignalbirdManagement:
    def __init__(
        self,
        domain_key: str,
        base_url: Optional[str] = None,
        timeout: float = 15.0,
        throw_on_error: bool = False,
        debug: bool = False,
    ):
        if not domain_key:
            raise SignalbirdError("Signalbird: domain_key zorunlu.", 0, "NO_KEY")

        if not domain_key.startswith("sb_secret_live_"):
            raise SignalbirdError(
                "Signalbird: yönetim istemcisi GİZLİ domain anahtarı ister (sb_secret_live_…).",
                0,
                "WRONG_KEY_TYPE",
            )

        self._http = Transport(domain_key, base_url or DEFAULT_BASE_URL, timeout, throw_on_error, debug)

    # ── Telsiz: projeler ─────────────────────────────────────────────────

    def radio_summary(self) -> Result:
        return self._http.request("GET", "/v1/radio/summary")

    def radio_events(self, query: Optional[Mapping[str, Any]] = None) -> Result:
        return self._http.request("GET", "/v1/radio/events", None, query)

    # ── Modül anahtarları ────────────────────────────────────────────────
    #
    # Telsiz projesi/kanalı ve uygulama kaydı 1 Eyl 2026'da kaldırıldı
    # (../signalbird.api/docs/KEY_ARCHITECTURE_2026-09-01.md §3). Beş modülün
    # (logger, email, sms, push, chat) hepsi aynı gövdeyi kullanır.

    def list_module_keys(self, module: str, query: Optional[Mapping[str, Any]] = None) -> Result:
        return self._http.request("GET", f"/v1/modules/{seg(module)}/keys", None, query)

    def get_module_key(self, module: str, id: Any) -> Result:
        return self._http.request("GET", f"/v1/modules/{seg(module)}/keys/{seg(id)}")

    def create_module_key(self, module: str, input: Mapping[str, Any]) -> Result:
        """``key`` verilmezse başlıktan üretilir; çakışırsa sonuna sayı eklenir."""
        return self._http.request("POST", f"/v1/modules/{seg(module)}/keys", input)

    def update_module_key(self, module: str, id: Any, input: Mapping[str, Any]) -> Result:
        """``key`` DEĞİŞTİRİLEBİLİR: eski ad 30 gün daha kabul edilir."""
        return self._http.request("PATCH", f"/v1/modules/{seg(module)}/keys/{seg(id)}", input)

    def delete_module_key(self, module: str, id: Any) -> Result:
        return self._http.request("DELETE", f"/v1/modules/{seg(module)}/keys/{seg(id)}")

    def list_module_key_devices(self, module: str, id: Any, query: Optional[Mapping[str, Any]] = None) -> Result:
        """Push kanalına kayıtlı cihazlar; token MASKELİ döner."""
        return self._http.request("GET", f"/v1/modules/{seg(module)}/keys/{seg(id)}/devices", None, query)

    # ── Sohbet: gelen kutusu ─────────────────────────────────────────────

    def chat_summary(self) -> Result:
        return self._http.request("GET", "/v1/chat/summary")

    def chat_updates(self) -> Result:
        return self._http.request("GET", "/v1/chat/updates")

    def list_conversations(self, query: Optional[Mapping[str, Any]] = None) -> Result:
        return self._http.request("GET", "/v1/chat/conversations", None, query)

    def get_conversation(self, id: str) -> Result:
        return self._http.request("GET", f"/v1/chat/conversations/{seg(id)}")

    def list_conversation_messages(self, id: str, query: Optional[Mapping[str, Any]] = None) -> Result:
        return self._http.request("GET", f"/v1/chat/conversations/{seg(id)}/messages", None, query)

    def start_conversation(self, input: Mapping[str, Any]) -> Result:
        """Proaktif sohbet — ziyaretçi yazmadan ajan başlatır."""
        return self._http.request("POST", "/v1/chat/conversations", input)

    def update_conversation(self, id: str, input: Mapping[str, Any]) -> Result:
        return self._http.request("PATCH", f"/v1/chat/conversations/{seg(id)}", input)

    def set_conversation_status(self, id: str, status: str) -> Result:
        return self._http.request("POST", f"/v1/chat/conversations/{seg(id)}/status", {"status": status})

    def assign_conversation(self, id: str, user_id: Optional[int] = None) -> Result:
        return self._http.request("POST", f"/v1/chat/conversations/{seg(id)}/assign", {"user_id": user_id})

    def read_conversation(self, id: str, last_message_id: Optional[str] = None) -> Result:
        return self._http.request(
            "POST", f"/v1/chat/conversations/{seg(id)}/read", {"last_message_id": last_message_id}
        )

    def set_typing(self, id: str, is_typing: bool) -> Result:
        return self._http.request(
            "POST", f"/v1/chat/conversations/{seg(id)}/typing", {"is_typing": is_typing}
        )

    def reply(self, id: str, input: Mapping[str, Any]) -> Result:
        return self._http.request("POST", f"/v1/chat/conversations/{seg(id)}/messages", input)

    def edit_chat_message(self, id: str, message_id: str, body: str) -> Result:
        return self._http.request(
            "PATCH", f"/v1/chat/conversations/{seg(id)}/messages/{seg(message_id)}", {"body": body}
        )

    def delete_chat_message(self, id: str, message_id: str) -> Result:
        return self._http.request(
            "DELETE", f"/v1/chat/conversations/{seg(id)}/messages/{seg(message_id)}"
        )

    def react_to_chat_message(self, id: str, message_id: str, emoji: str) -> Result:
        """Aynı emoji ikinci kez gönderilirse tepki kaldırılır."""
        return self._http.request(
            "POST",
            f"/v1/chat/conversations/{seg(id)}/messages/{seg(message_id)}/reactions",
            {"emoji": emoji},
        )

    # ── Sohbet: ziyaretçi ve hazır yanıtlar ──────────────────────────────

    def get_visitor(self, id: str) -> Result:
        return self._http.request("GET", f"/v1/chat/visitors/{seg(id)}")

    def update_visitor(self, id: str, input: Mapping[str, Any]) -> Result:
        return self._http.request("PATCH", f"/v1/chat/visitors/{seg(id)}", input)

    def ban_visitor(self, id: str) -> Result:
        return self._http.request("POST", f"/v1/chat/visitors/{seg(id)}/ban")

    def list_canned_replies(self) -> Result:
        return self._http.request("GET", "/v1/chat/canned-replies")

    def create_canned_reply(self, input: Mapping[str, Any]) -> Result:
        return self._http.request("POST", "/v1/chat/canned-replies", input)

    def update_canned_reply(self, id: Any, input: Mapping[str, Any]) -> Result:
        return self._http.request("PATCH", f"/v1/chat/canned-replies/{seg(id)}", input)

    def delete_canned_reply(self, id: Any) -> Result:
        return self._http.request("DELETE", f"/v1/chat/canned-replies/{seg(id)}")

    # ── Sohbet: tetikleyiciler ───────────────────────────────────────────
    # "Şu olduğunda şunu yap." Kural KAYITTA durur, kodda değil.

    def list_chat_triggers(self) -> Result:
        return self._http.request("GET", "/v1/chat/triggers")

    def create_chat_trigger(self, input: Mapping[str, Any]) -> Result:
        return self._http.request("POST", "/v1/chat/triggers", input)

    def update_chat_trigger(self, id: Any, input: Mapping[str, Any]) -> Result:
        return self._http.request("PATCH", f"/v1/chat/triggers/{seg(id)}", input)

    def delete_chat_trigger(self, id: Any) -> Result:
        return self._http.request("DELETE", f"/v1/chat/triggers/{seg(id)}")

    # ── Sohbet: rapor ────────────────────────────────────────────────────

    def chat_report(self, range: str = "30d") -> Result:
        """Yanıt/çözüm süresi, memnuniyet, ajan kırılımı.

        Veri yoksa süreler ``None`` döner — 0 DEĞİL.
        """
        return self._http.request("GET", "/v1/chat/reports", None, {"range": range})

    # ── Uygulamalar ──────────────────────────────────────────────────────

    # Uygulama uçları KALDIRILDI (1 Eyl 2026): sohbet ve push birer modül
    # anahtarıdır — list_module_keys('chat'), list_module_keys('push').

    def embed_token(self, input: Mapping[str, Any]) -> Result:
        """Gömme jetonu — Signalbird ekranını kendi panelinizde göstermek için.

        120 saniye yaşar ve tek kullanımlıktır; anahtar `embed:issue` kapsamı ister.
        """
        return self._http.request("POST", "/v1/embed/tokens", dict(input))

